project/src/pv_circular_buffer.c

```diff
@@ -0,0 +1,170 @@
+/*
+    Copyright 2024 Picovoice Inc.
+
+    You may not use this file except in compliance with the license. A copy of the license is located in the "LICENSE"
+    file accompanying this source.
+
+    Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+    an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+    specific language governing permissions and limitations under the License.
+*/
+
+#include <stdlib.h>
+#include <string.h>
+
+#include "pv_circular_buffer.h"
+
+struct pv_circular_buffer {
+    void *buffer;
+    int32_t capacity;
+    int32_t count;
+    int32_t element_size;
+    int32_t read_index;
+    int32_t write_index;
+};
+
+pv_circular_buffer_status_t pv_circular_buffer_init(
+        int32_t element_count,
+        int32_t element_size,
+        pv_circular_buffer_t **object) {
+    if (element_count <= 0) {
+        return PV_CIRCULAR_BUFFER_STATUS_INVALID_ARGUMENT;
+    }
+    if (element_size <= 0) {
+        return PV_CIRCULAR_BUFFER_STATUS_INVALID_ARGUMENT;
+    }
+    if (!object) {
+        return PV_CIRCULAR_BUFFER_STATUS_INVALID_ARGUMENT;
+    }
+
+    *object = NULL;
+
+    pv_circular_buffer_t *o = calloc(1, sizeof(pv_circular_buffer_t));
+    if (!o) {
+        return PV_CIRCULAR_BUFFER_STATUS_OUT_OF_MEMORY;
+    }
+
+    o->buffer = malloc(element_count * element_size);
+    if (!(o->buffer)) {
+        pv_circular_buffer_delete(o);
+        return PV_CIRCULAR_BUFFER_STATUS_OUT_OF_MEMORY;
+    }
+
+    o->capacity = element_count;
+    o->element_size = element_size;
+
+    *object = o;
+
+    return PV_CIRCULAR_BUFFER_STATUS_SUCCESS;
+}
+
+void pv_circular_buffer_delete(pv_circular_buffer_t *object) {
+    if (object) {
+        free(object->buffer);
+        free(object);
+    }
+}
+
+pv_circular_buffer_status_t pv_circular_buffer_read(
+        pv_circular_buffer_t *object,
+        void *buffer,
+        int32_t buffer_length) {
+    if (!object) {
+        return PV_CIRCULAR_BUFFER_STATUS_INVALID_ARGUMENT;
+    }
+    if (!buffer) {
+        return PV_CIRCULAR_BUFFER_STATUS_INVALID_ARGUMENT;
+    }
+    if (!buffer_length) {
+        return PV_CIRCULAR_BUFFER_STATUS_INVALID_ARGUMENT;
+    }
+    if ((buffer_length <= 0) || (buffer_length > object->capacity)) {
+        return PV_CIRCULAR_BUFFER_STATUS_INVALID_ARGUMENT;
+    }
+
+    void *dst_ptr = buffer;
+    const void *src_ptr = (char *) object->buffer + (object->read_index * object->element_size);
+
+    const int32_t available = object->capacity - object->read_index;
+    const int32_t max_copy = (object->count < buffer_length) ? object->count : buffer_length;
```

Comment:
so we leave the rest of the buffer untouched without notifying the user?

Reply:
the rest of the buffer is also copied in the `if` block below